File: 线性表/静态链表/StaticList.java
@SuppressWarnings("unchecked")
public class StaticList<T> {
    public static class Node<T> {
        private T data;
        private int cursor;

        public Node(T data, int cursor) {
            this.data = data;
            this.cursor = cursor;
        }

        private void setCursor(int cursor) {
            this.cursor = cursor;
        }

        public int getCursor() {
            return cursor;
        }

        private void setData(T data) {
            this.data = data;
        }

        public T getData() {
            return data;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "data=" + data +
                    ", cursor=" + cursor +
                    '}';
        }
    }

    private Node<T>[] underlying;
    private Node<T> head, space;

    public StaticList(int totalLength) {
        if (totalLength < 3)
            throw new RuntimeException("totalLength < 3");

        underlying = (Node<T>[]) new Node<?>[totalLength];
        underlying[0] = space = new Node<T>(null, 0);
        int i = 0;
        for (; i < totalLength - 1; i++) {
            underlying[i + 1] = new Node<T>(null, 0);
            underlying[i].setCursor(i + 1);
        }
        underlying[i].setCursor(0);
        head = underlying[malloc()];
        head.setCursor(0);
    }

    public T remove(T data) {
        if (data == null)
            throw new RuntimeException("data == null");

        Node<T> node = head, nextNode;
        int nextIndex, nextNodeIndex;
        while (node.getCursor() != 0) {
            nextNode = underlying[node.getCursor()];
            if (data.equals(nextNode.getData())) {
                nextIndex = nextNode.getCursor();
                nextNodeIndex = node.getCursor();
                free(nextNodeIndex);
                node.setCursor(nextIndex);
                return data;
            }
            node = underlying[node.getCursor()];
        }
        return null;
    }

    public boolean add(T data) {
        if (data == null)
            throw new RuntimeException("data == null");

        int index = malloc();
        // 没有空闲节点
        if (index == 0)
            return false;

        underlying[index].setData(data);
        underlying[index].setCursor(0);

        Node<T> node = head;
        while (node.getCursor() != 0)
            node = underlying[node.getCursor()];
        node.setCursor(index);
        return true;
    }

    // 将index放到备用链表表头
    private void free(int index) {
        underlying[index].setCursor(space.getCursor());
        space.setCursor(index);
    }

    // 从备用链表表头获取空闲节点
    private int malloc() {
        int index;
        if ((index = space.getCursor()) == 0)
            return 0;
        space.setCursor(underlying[index].getCursor());
        return index;
    }

    public static void main(String[] args) {
        StaticList<Integer> staticList = new StaticList<Integer>(4);
        System.out.println(staticList.add(1));
        System.out.println(staticList.add(2));
        System.out.println(staticList.add(3));
        System.out.println(staticList.remove(2));
        System.out.println(staticList.remove(333));
        System.out.println(staticList.add(3));
        System.out.println(staticList.remove(1));
        System.out.println(staticList.add(2));
        System.out.println(staticList.remove(3));
        int index = 0;
        for (Node<Integer> element: staticList.underlying)
            System.out.println("index = " + index++ + ", element = " + element);
    }
}

File: 线性表/静态链表/static_list.py
# coding: utf8


class Node(object):
    """
    静态链表的节点
    """
    def __init__(self, element, cursor=-1):
        # 保存数据元素
        self.element = element
        # 保存直接后继节点在数组中的索引
        self.cursor = cursor


class StaticList(object):
    """
    静态链表
    """
    def __init__(self, max_size):
        # 备用链表的头节点
        self._free_space = Node(None, -1)
        # 静态链表的头节点
        self._head = Node(None, -1)
        # 底层数组
        self._underlying_array = [Node(None, -1) for _ in range(max_size)]

        # 将数组中的全部节点组织进备用链表
        self._free_space.cursor = 0
        for ind in range(0, max_size - 1):
            self._underlying_array[ind].cursor = ind + 1

    def _malloc(self):
        """
        用于从备用链表中申请节点
        """
        cursor = self._free_space.cursor
        if cursor == -1:
            raise RuntimeError("no space left")
        node = self._underlying_array[cursor]
        self._free_space.cursor = node.cursor
        return cursor

    def _free(self, cursor):
        """
        将节点放回备用链表
        """
        node = self._underlying_array[cursor]
        node.cursor = self._free_space.cursor
        self._free_space.cursor = cursor

    def insert(self, index, element):
        node = self._head
        current_index = -1

        while node is not None:
            if current_index == index - 1:
                # 申请新节点
                free_cursor = self._malloc()
                free_node = self._underlying_array[free_cursor]
                free_node.element = element
                free_node.cursor = node.cursor
                node.cursor = free_cursor
                break

            if node.cursor == -1:
                node = None
                continue

            node = self._underlying_array[node.cursor]
            current_index = current_index + 1
        else:
            raise IndexError("invalid index")

    def find(self, element):
        temp = self._head.cursor
        index = 0

        while temp != -1:
            node = self._underlying_array[temp]
            if node.element == element:
                return index
            temp = node.cursor
            index = index + 1
        else:
            raise ValueError("not found")

    def delete(self, element):
        # node 是待删除节点的前一个节点
        node = self._head
        if node.cursor == -1:
            next_node = None
        else:
            next_node = self._underlying_array[node.cursor]

        while next_node is not None:
            if next_node.element == element:
                next_node_cursor = next_node.cursor
                # 释放节点
                self._free(node.cursor)
                # 删除节点
                node.cursor = next_node_cursor
                break
            node = next_node
            if node.cursor == -1:
                next_node = None
            else:
                next_node = self._underlying_array[node.cursor]
        else:
            raise ValueError("not found")


if __name__ == "__main__":
    import unittest

    class TestStaticList(unittest.TestCase):
        def testStaticList(self):
            static_list = StaticList(5)
            static_list.insert(0, 1)
            static_list.insert(0, 0)
            static_list.insert(2, 2)
            self.assertRaises(IndexError, static_list.insert, 4, 4)
            self.assertRaises(ValueError, static_list.delete, 4)

            self.assertEqual(static_list.find(0), 0)
            self.assertEqual(static_list.find(1), 1)
            self.assertEqual(static_list.find(2), 2)
            self.assertRaises(ValueError, static_list.find, -1)

            static_list.delete(1)
            self.assertEqual(static_list.find(2), 1)
            static_list.delete(0)
            self.assertEqual(static_list.find(2), 0)

    unittest.main()
